Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -2371,6 +2371,10 @@ protected class ClassWriter : IClassWriter {
       to = to.NormalizeExpand();
       Contract.Assert(from.IsRefType == to.IsRefType);
 
+      if (from.IsIndDatatype || from.IsCoDatatype) {
+        Error(tok, "compilation does not support downcasts involving copying for datatypes (like converting from {0} to {1})", wr, from, to);
```

Comment:
Could you add some information why the C# code currently doesn't compile? I was hoping a single cast in C# would be enough to make this work, like 

```
List<Vehicle> vehicle = new Car();
List<Car> car = (List<Car>)vehicle;
```

but apparently it doesn't..?

Reply:
Without the check, a call to DowncastClone of the datatype is generated, but we haven’t implemented it. We have some handwritten ones in the DafnyRuntime.cs, if you want to get an idea of that they do.

Reply:
Right, but do we want to call DowncastClone instead casting? What happens if you change `if (to.IsRefType)` into `if (to.isRefType || to.IsIndDatatype || to.IsCoDatatype)` ?

Reply:
Thanks for the suggestion! It works for my test, but I'll try to figure out, if it's a general solution.

Reply:
Here's a counterexample:
```
datatype Co<+T> = Co(T)

trait Number {
  const value: int
}

class Integer extends Number {
  constructor(value: int) {
    this.value := value;
  }
}

method Main(){
  var x := new Integer(20);
  var y := new Integer(30);
  var s: seq<Number>;
  s := [x, y];

  var a: Co<seq<Number>> := Co(s);
  var b: Co<seq<Integer>>;
  b := a;
}
```

Reply:
Could you add more information? What happens for this program?

Seems like `seq` will also return `false` for `IsRefType` so it doesn't generate a cast for `b := a`, even though it should.

Reply:
What I meant was, that with your suggested change, I can emit a cast like the one you suggested, but it fails at runtime. In this case we really need a `DowncastClone`.

Reply:
Why does it fail at runtime? Doesn't the verification ensure that the target type equals the actual type?

Reply:
I think that the problem is a mismatch between what Dafny can prove to be possible and what a C# cast is willing to do.

Reply:
Can you paste the runtime error?

Reply:
```
Unhandled exception. System.InvalidCastException: Unable to cast object of type '_module.Co`1[_module.Number]' to type '_module._ICo`1[_module.Integer]'.
   at _module.__default._Main()
   at Dafny.Helpers.WithHaltHandling(Action action)
   at __CallToMain.Main(String[] args)
```

Reply:
Changing the target to `_module.Co1[_module.Integer]` doesn't help either.